### ISSUE TITLE
GCP ConfigConnector CRD - No valid OpenAPI definition Error

### ISSUE THREAD

**Issue by kaeyee** [hashicorp/terraform-provider-kubernetes-alpha#181]
Hello, we are facing a OpenAPI definition error after upgrading our provider from v0.2.1 to v0.3.1.
We are making use of kubernetes-alpha provider to apply the GCP ConfigConnector, specifically for Redis instance.
As checked in the GKE CRD, the redisinstances.redis.cnrm.cloud.google.com CRD has schema definition.

### Terraform, Provider, Kubernetes versions

```
Terraform version: v0.14.8
Provider version: 0.3.2
Kubernetes version: v1.18.15-gke.1501
```

### Affected Resource(s)

kubernetes_manifest

### Terraform Configuration Files
```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
resource "kubernetes_manifest" "redis_cr" {
  provider = kubernetes-alpha

  manifest = {
    apiVersion = "redis.cnrm.cloud.google.com/v1beta1"
    kind       = "RedisInstance"
    metadata = {
      name      = local.redis_name
      namespace = var.namespace
      labels    = var.labels
      annotations = {
        "cnrm.cloud.google.com/project-id" = var.project
      }
    }
    spec = {
      memorySizeGb          = var.memory_size_gb
      region                = var.region
      tier                  = var.tier
      redisVersion          = var.redis_version
      alternativeLocationId = var.alternative_location_id
      authEnabled           = var.auth_enabled
      connectMode           = var.connect_mode
      displayName           = var.display_name
      locationId            = var.location_id
      reservedIpRange       = var.reserved_ip_range
      redisConfigs          = var.redis_configs
      authorizedNetworkRef  = var.authorized_network_ref
    }
  }

  wait_for = {
    fields = {
      # Check the phase of redis cr
      "status.conditions[0].type" = "Ready"

      # Check the status of a redis cr
      "status.conditions[0].status" = "True"
    }
  }
}
```

### Debug Output

![image](https://user-images.githubusercontent.com/9191428/112285934-cbc40800-8cc5-11eb-80f3-5de9eb0ad0dc.png)

### Steps to Reproduce

1. terraform apply

### Expected Behavior
What should have happened?
apply successfully pass

### Actual Behavior
What actually happened?
Error thrown, as shown in the attached picture.

### References

- GH-1234

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by Vrtak-CZ** [comment]
Hi, i have same problem with `ExternalSecret`.

Provider returns `Error: No valid OpenAPI definition` even when CRD has OpenApi Schema https://github.com/external-secrets/kubernetes-external-secrets/blob/master/charts/kubernetes-external-secrets/crds/kubernetes-client.io_externalsecrets_crd.yaml#L33-L133

**Comment by flolu** [comment]
Same problem with `kafka.strimzi.io/v1beta2`

**Comment by brandondtb** [comment]
I am getting the same error with ` ManagedCertificate.v1.networking.gke.io`.

**Comment by dplacek** [comment]
Also seeing this for `Mapping.v2.getambassador.io`.

**Comment by voatsap** [comment]
same error with argocd apps! 
Resource Application.v1alpha1.argoproj.io does not have a valid OpenAPI
definition in this cluster

**Comment by dniel** [comment]
and same for Helm "Resource HelmRelease.v1.helm.fluxcd.io does not have a valid OpenAPI definition in this cluster"

**Comment by so0k** [comment]
same error for https://github.com/isindir/sops-secrets-operator/blob/master/config/crd/bases/isindir.github.com_sopssecrets.yaml#L27

**Comment by vauchok** [comment]
I'm also faced the issue:
`Resource IngressGateway.v1alpha1.consul.hashicorp.com does not have a valid `

```
Terraform version: v0.14.9
Provider version: 0.3.2
Kubernetes version: v1.18.9-eks-d1db3c
```

**Comment by TrevorPace** [comment]
This fails during the planning stage even though the resource is already available in cluster.

I find this a bit frustrating as https://github.com/hashicorp/terraform-provider-kubernetes-alpha/issues/72 was closed supposedly because the changes in 0.3.0 should have fixed it, yet clearly no one tested that it actually worked. Seeing as the current production kubernetes provider already defines all of the built-in kubernetes resources I question what the use of this alpha-provider even is if it doesn't actually work with CRDs properly.

**Comment by alexsomesan** [comment]
Hi all,

Sorry for the late reply to this issue. I've had a look at it and tried to reproduce with GKE. It turns out that indeed, the RedisInstance CRs cannot be created even when their parent CRD is present in the cluster.
The technical explanation for this is that, for some strange reason, even if the CRD is installed the cluster OpenAPI spec at `/openapi/v2` does not reflect any of the schema defined by the CRD for RedisInstance so the provider cannot determine the schema for it.

Long story short, we already have a couple of reports of similar issues and have looked into solutions. We plan to try to solve this by fetching the schema information off the CRD objects directly, rather than through the cluster OpenAPI. We're hoping to prioritise this work in the next few weeks.

**Comment by rjcoupe** [comment]
+1, getting this with `Application.v1alpha1.argoproj.io`.

**Comment by deserializeme** [comment]
+1

Error
```
Error: No valid OpenAPI definition

  on .terraform/modules/pod_identity_tests/resources.tf line 82, in resource "kubernetes_manifest" "azure_identity_resource":
  82: resource "kubernetes_manifest" "azure_identity_resource" {

Resource AzureIdentityBinding.v1.aadpodidentity.k8s.io does not have a valid
OpenAPI definition in this cluster.

Usually this is caused by a CustomResource without a schema.
```

resource:
```
resource "kubernetes_manifest" "azure_binding_resource" {
  provider = "kubernetes-alpha"
  manifest = {
    "apiVersion" = "aadpodidentity.k8s.io/v1"
    "kind"       = "AzureIdentity"
    "metadata" = {
      "name"      = "${module.aad_demo_user_identity.name}"
      "namespace" = "${var.namespace}"
    }
    "spec" = {
      "clientID"   = "${module.aad_demo_user_identity.client_id}"
      "resourceID" = "${module.aad_demo_user_identity.id}"
      "type"       = 0
    }
  }
}
```

provider:
```
    kubernetes-alpha = {
      source = "hashicorp/kubernetes-alpha"
      version = "~>0.3.2"
    }
```

azure AKS k8s v 1.19.7 on ubuntu 18.04

**Comment by alexsomesan** [comment]
Hi all,

I did a bit more digging here on the original issue on GKE. It turns out that in the specific case of a GKE cluster with addons enabled, the provider is having difficulties retrieving the CRD resource via the discovery API. This seems specifically related to add-ons being turned on. A vanilla GKE cluster will behave correctly.

I'm continuing to debug this to figure out what's happening in the discovery client when add-ons are turned on. I'll update with more info shortly.

**Comment by revog** [comment]
Got same issue with on-prem RKE2 (rancher) k8s cluster.

`Terraform version: v0.15.0
Provider version: v0.3.2
Kubernetes version: v1.20.4+rke2r1`

Error:
`¦ Error: No valid OpenAPI definition
¦
¦   on ../modules/rancher/backup.tf line 87, in resource "kubernetes_manifest" "hourly-backup":
¦   87: resource "kubernetes_manifest" "hourly-backup" {
¦
¦ Resource Backup.v1.resources.cattle.io does not have a valid OpenAPI definition in this cluster.
¦
¦ Usually this is caused by a CustomResource without a schema.
`
Resource:
`resource "kubernetes_manifest" "hourly-backup" {
  provider = kubernetes-alpha

  manifest = {
    apiVersion = "resources.cattle.io/v1"
    kind = "Backup"
    metadata = {
      name = "hourly-backup-s3"
    }
    spec = {
      encryptionConfigSecretName: kubernetes_secret.encryption
      schedule: "@every 1h"
      storagelocation = {
        s3 = {
          bucketName = "rancher2-etcd"
          credentialSecretName = kubernetes_secret.s3
          credentialSecretNamespace = var.rancher_namespace
          endpoint = "localhost:9000"
          folder = var.stage
          insecureTLSSkipVerify = true
        }
      }
      resourceSetName: "rancher-resource-set"
      retentionCount: 36
    }
  }
}`

**Comment by TrevorPace** [comment]
> Hi all,
> 
> I did a bit more digging here on the original issue on GKE. It turns out that in the specific case of a GKE cluster with addons enabled, the provider is having difficulties retrieving the CRD resource via the discovery API. This seems specifically related to add-ons being turned on. A vanilla GKE cluster will behave correctly.
> 
> I'm continuing to debug this to figure out what's happening in the discovery client when add-ons are turned on. I'll update with more info shortly.

This is also broken on AWS as well as a heads up.

**Comment by bhagtrajaram** [comment]
hi,

Getting the same error as mentioned above:

kubernetes_manifest.test-namespace: Refreshing state...

Error: No valid OpenAPI definition

  on main.tf line 83, in resource "kubernetes_manifest" "template-allow-privilege-escalation-constraint":
  83: resource "kubernetes_manifest" "template-allow-privilege-escalation-constraint" {

Resource ConstraintTemplate.v1beta1.templates.gatekeeper.sh does not have a
valid OpenAPI definition in this cluster.

Usually this is caused by a CustomResource without a schema.

ERRO[0021] Hit multiple errors:
Hit multiple errors:
exit status 1

**Comment by ivanov-aleksander** [comment]
I am also have the same issue

**Comment by ivanov-aleksander** [comment]
In my case i have such kubernetes_manifest resource
```
resource "kubernetes_manifest" "vertical-pod-autoscaler" {
  provider = kubernetes-alpha

  manifest = {
    "apiVersion" = "autoscaling.k8s.io/v1"
    "kind"       = "VerticalPodAutoscaler"
     ...
```
Terraform plan fails with error
```
╷
│ Error: No valid OpenAPI definition
│ 
│   with kubernetes_manifest.vertical-pod-autoscaler,
│   on vpa.tf line 4, in resource "kubernetes_manifest" "vertical-pod-autoscaler":
│    4: resource "kubernetes_manifest" "vertical-pod-autoscaler" {
│ 
│ Resource VerticalPodAutoscaler.v1.autoscaling.k8s.io does not have a valid OpenAPI definition in this cluster.
│ 
│ Usually this is caused by a CustomResource without a schema.
```

Cluster has verticalpodautoscalers api resource with Kind: VerticalPodAutoscaler
```bash
kubectl api-resources --api-group=autoscaling.k8s.io   
NAME                               SHORTNAMES      APIVERSION              NAMESPACED   KIND
verticalpodautoscalers             vpa             autoscaling.k8s.io/v1   true         VerticalPodAutoscaler
```
It looks like kubernetes-alpha checks _Kind_ instead of CRD. In my exaple it checks _Resource VerticalPodAutoscaler.v1.autoscaling.k8s.io_ instead of _Resource VerticalPodAutoscaler**s**.v1.autoscaling.k8s.io_

**Comment by larryboymi** [comment]
Can confirm same issue exists trying to use traefik CRDs

**Comment by puzzloholic** [comment]
Same error on Istio Operator 1.9.5

```
│ Error: No valid OpenAPI definition
│
│   with kubernetes_manifest.istio_operator,
│   on istio.tf line 15, in resource "kubernetes_manifest" "istio_operator":
│   15: resource "kubernetes_manifest" "istio_operator" {
│
│ Resource IstioOperator.v1alpha1.install.istio.io does not have a valid OpenAPI definition in this cluster.
│
│ Usually this is caused by a CustomResource without a schema.
```

```
Terraform v0.15.3
on linux_amd64

provider:
hashicorp/kubernetes-alpha 0.3.2
```

**Comment by LiuVII** [comment]
Same here, can't use `kubernetes_manifest` on versions 0.3.0+ with Istio Operator/Kiali/Jaeger.
Any updates/workarounds?

**Comment by puzzloholic** [comment]
> Same here, can't use `kubernetes_manifest` on versions 0.3.0+ with Istio Operator/Kiali/Jaeger.
> Any updates/workarounds?

@LiuVII I use `gavinbunney/kubectl` provider for manifests failing on this provider

**Comment by LiuVII** [comment]
Recent 0.4.1 version fixed the issue for me which was support of both Istio Operator and Prom Operator stacks

**Comment by deserializeme** [comment]
Thanks for updating the thread, I'll check my own setup on azure again soon and report back

**Comment by alexsomesan** [comment]
We've just released version 0.5.0 which contains improvements to CRD handling. 
Would you all mind trying this new version out with your use-cases and sharing your results?

Thank you!

**Comment by alexsomesan** [comment]
@Vrtak-CZ thanks for confirming ! 

A bit of clarification there: you're getting that warning in 0.4.1 because the schema for the CRD cannot be found in the cluster-wide OpenAPI spec (on the API server at URL path `/openapi/v2`). The warning is not about the OpenAPI schema embedded in the CRD itself. This usually happens when CRDs are either of version `apiextensions.k8s.io/v1beta1` or they have `preserveUnknownFields` turned on explicitly. In that case, the API will not publish the OpenAPI schema anymore. 

To work around that, in provider version 0.5.0 we've implemented a mechanism to extract the schema directly from the CRD object itself. This should be more reliable and you should also not see that warning anymore, unless the CRD really doesn't have any schema.

**Comment by dniel** [comment]
Hi, I have testet version 0.5.0 now and tried to deploy external-secrets using the HelmRelease CRD. Previously this code failed on Missing OpenAPI schema, but now it fails on something new. 

Terraform output
```
When applying changes to
module.template-services.module.secrets.kubernetes_manifest.external-secrets-helm-release,
provider "registry.terraform.io/hashicorp/kubernetes-alpha" produced an
unexpected new value: .object: wrong final value type: incorrect object
attributes.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

```

The manifest is here https://github.com/dniel/terraform/blob/745f508b79d501ddb1c209f1ba66e2ced7264eb1/modules/template-services/secrets/main.tf#L48-L86

**Comment by Vrtak-CZ** [comment]
> To work around that, in provider version 0.5.0 we've implemented a mechanism to extract the schema directly from the CRD object itself. This should be more reliable and you should also not see that warning anymore, unless the CRD really doesn't have any schema.

Yes that's what I expected but I see this waring on 0.5.0 too (but CRD has schema defined).

**Comment by alexsomesan** [comment]
I'm going to close this since we can now reliably confirm the fix for the original issue report (and others mentioned in the comments).

If anyone is still experiencing issues similar to this I would encourage you to open specific issues with examples or instructions to reproduce. It's easier to handle each report in its own context.

**Comment by dniel** [comment]
I bit early to close? 
I cant see any feedback yet that anyone actually verified it successfully?
